### ISSUE TITLE
Opt-Out not working

### ISSUE THREAD

**Issue by stillworks** [insites/cookieconsent#452]
Ideally if I chhose opt-out, all my tracking should be load during the first page load. It won't. I stays as a opt-in kinda. I tried to fiddle around where In and Out scripts are placed, but no success.

Opt-In works and just the notification.

```javascript
window.addEventListener("load", function(){
  window.cookieconsent.initialise({
    "palette": {
      "popup": {
        "background": "#000",
        "text": '#fff'
      },
      "button": {
        "background": '#f29400',
        "text": '#fff'
      }
    },
    "type": "opt-out",
    "revokable": true,
    "content": {
      "message": ''      ,
      "dismiss": 'Alles klar',
      "deny": 'no',
      "allow": 'Bok',
      "link": 'Zur Datenschutzerklärung',
      "href": '/datenschutz',
      "policy": '<i class="fas fa-cookie-bite"></i> Cookie Einstellung',
    },

    onInitialise: function (status) {
      var type = this.options.type;
      var didConsent = this.hasConsented();
      if (type == 'opt-in' && didConsent) {
        // enable cookies
        trackingScriptsIn();

      }
      if (type == 'opt-out' && !didConsent) {
        // disable cookies
        trackingScriptsIn();

      }
    },

    onStatusChange: function(status, chosenBefore) {
      var type = this.options.type;
      var didConsent = this.hasConsented();
      if (type == 'opt-in' && didConsent) {
        // enable cookies
        trackingScriptsIn();


      }
      if (type == 'opt-out' && !didConsent) {
        // disable cookies
        trackingScriptsOut();

      }
      window.location.href = window.location.href;
    },

    onRevokeChoice: function() {
      var type = this.options.type;
      if (type == 'opt-in') {
        // disable cookies
        trackingScriptsOut();


      }
      if (type == 'opt-out') {
        // enable cookies
        trackingScriptsIn();

      }
    }
  })});
```

**Comment by thexmanxyz** [comment]
@stillworks I didn't really try to reproduce your code but I checked it roughly. The approach looks correctly but I assume there is an issue with your conditions. This is just an assumption as I haven't replayed anything but to me the condition do not cover the actual status correctly. Use developer tools to debug your code on initialization this should show up any discrepancies.

**Comment by webjunkie** [comment]
Please see https://github.com/insites/cookieconsent/issues/352

You have to not rely on `onInitialised`.

**Comment by thexmanxyz** [comment]
@webjunkie Good point and reference this should help the issue starter @stillworks a lot!

**Comment by Chuvisco88** [comment]
@stillworks I had the same issue and i fixed it manually, se below code:
`function getCookie(name) {
        var b = document.cookie.match('(^|[^;]+)\\s*' + name + '\\s*=\\s*([^;]+)');
        return b ? b.pop() : '';
    }

    window.addEventListener("load", function () {
        const cookieConsent = getCookie('cookieconsent_status');

        if (cookieConsent === '' || cookieConsent === 'allow') {
            console.log('Load cookies');
        } else {
            console.log('No having cookies');
        }

        window.cookieconsent.initialise({
            "palette": {
                "popup": {
                    "background": "#efefef",
                    "text": "#404040"
                },
                "button": {
                    "background": "#8ec760",
                    "text": "#ffffff"
                }
            },
            "type": "opt-out",
            "content": {
                "allow": "Approve",
                "dismiss": "Approve",
                "deny": "Reject"
            },
            onStatusChange: function (status, chosenBefore) {
                const type = this.options.type;
                const didConsent = this.hasConsented();

                location.reload();
            }
        })
    });`

**Comment by relicmelex** [comment]
@stillworks are you still experiencing this issue?

**Comment by iamwebrocker** [comment]
Hi, anyone else noticed that `this.hasConsented()`always returns "true", no matter which button the user pressed (dismiss | allow).
I had to check for "status" as well to get the right opt-in allow / dismiss behavior (onyl shown for the opt-in type here):

```
onInitialise: function (status) {
		var type = this.options.type;
		var didConsent = this.hasConsented();

		if (type == 'opt-in' && didConsent) {
			if (status == 'allow'){
				_paq.push(['rememberConsentGiven']);
			}
			if (status == 'dismiss'){
				_paq.push(['forgetConsentGiven']);
			}
		}
		if (type == 'opt-out' && !didConsent) {
			// disable cookies
		}
	},
	onStatusChange: function(status, chosenBefore) {
		var type = this.options.type;
		var didConsent = this.hasConsented();
		if (type == 'opt-in' && didConsent) {
			// enable cookies
			if (status == 'allow'){
				_paq.push(['rememberConsentGiven']);
			}
			if (status == 'dismiss'){
				_paq.push(['forgetConsentGiven']);
			}
		}
		if (type == 'opt-out' && !didConsent) {
			// disable cookies
		}
	},
	onRevokeChoice: function() {
		var type = this.options.type;
		if (type == 'opt-in') {
			// disable cookies
			_paq.push(['forgetConsentGiven']);
		}
		if (type == 'opt-out') {
			// enable cookies
		}
	}
```
Dunno if that's how it is supposed to work, it definetly puzzled me. My expectation was that this.hasConsented would return true / false, depending on what the user clicked.
I guess it makes sense to have this in the status, and the "hasConsented" as a signal if the user has interacted with the message in any way, but then the documentation on allowing/deleting cookies is a bit misleading, for me at least :-)

**Comment by thexmanxyz** [comment]
@iamwebrocker As you probably already found out yourself `hasConsented(...)` does not indicate if the permission was give or not. See the code below:

```JS
// returns true if the cookie indicates that consent has been given
CookiePopup.prototype.hasConsented = function(options) {
  var val = this.getStatus();
  return val == cc.status.allow || val == cc.status.dismiss;
};
```

It only check whether there is a cookie status set. If the status is `dismiss` or `allow` doesn't matter. To sum things up, if there is no user action stored in the cookie (neither allow or dismiss) it returns `false` otherwise `true`. I'm pretty sure it is intended this way. I tag @relicmelex to look after this because a overall rework is planed.

**Comment by iamwebrocker** [comment]
@thexmanxyz thank you for the clarification / confirmation. :)

**Comment by relicmelex** [comment]
Thanks for tagging me. Yes I noticed the same thing and thought it strange. Honestly it’s just poor practice to have a config function use the scope of the generated object based on the config to get information... I will look into this more soon and follow up with my recommendations for how to leverage this while still in v3. v4 and the following versions will do this slightly differently anyways ( probably an event or just a initialize callback parameter )

**Comment by iamwebrocker** [comment]
@thexmanxyz 
> In the meantime please tell us what part of docs you found misleading. Thanks.

ah, sorry, forgot to answer this in the last reply.
On here: https://cookieconsent.osano.com/documentation/disabling-cookies/

```
onInitialise: function (status) {
 var type = this.options.type;
 var didConsent = this.hasConsented();
 if (type == 'opt-in' && didConsent) {
  // enable cookies
 }
 if (type == 'opt-out' && !didConsent) {
  // disable cookies
 }
},
onStatusChange: function(status, chosenBefore) {
 var type = this.options.type;
 var didConsent = this.hasConsented();
 if (type == 'opt-in' && didConsent) {
  // enable cookies
 }
 if (type == 'opt-out' && !didConsent) {
  // disable cookies
 }
},
onRevokeChoice: function() {
 var type = this.options.type;
 if (type == 'opt-in') {
  // disable cookies
 }
 if (type == 'opt-out') {
  // enable cookies
 }
}
```

There is no check for the allow/dismiss status in the example code.
Just the code comments for enable/disable cookies at the corresponding places where `has.consented` is involved - this implied to me that `has.consented` would carry the true/false signal regarding the status, as in 'user has given consent', aka allows cookies.
But since a user can give consent - i.e. has interacted with the banner, but could also dismiss, enabling cookies just based on `has.consented` as it is implied by the code comment in the section for opt-in for example, is not correct, and a 'dismiss' user would allow cookie storing.

**Comment by thexmanxyz** [comment]
@iamwebrocker I reference this back on the documentation update issue #568 and tag @relicmelex again. I do understand now, I didn't see that in the docs. I checked something else. Thanks for providing this useful additional information. This will help when updating the docs! This is indeed very misleading in that case because the `status` is not involved at all when it should be.

**Comment by iamwebrocker** [comment]
@thexmanxyz :-) Glad to be of help. I edit the above post and add my (works-for-me) code. Thanks for looking into this (and btw thank you all for the cookieconsent thing, this is very helpful)

**Comment by thexmanxyz** [comment]
@iamwebrocker You are very welcome! Thanks for the comprehensive explanation on how it really should work and be documented.

**Comment by iamwebrocker** [comment]
@thexmanxyz :-) while we are at it… to me the difference between the opt-in and opt-out banner, in regard of the function of the buttons (and thus the status involved) is a bit unclear. The "allow" button is used to "dismiss" the message, but b/c opt-out, allow cookies, right? And the "let me opt out" button is the real "dismiss" button? So in both cases opt-out and opt-in we'd have the same buttons with the same status - and initial behavior aside - there really isn't a difference wether it is an opt-out or opt-in banner, in both cases we need to check with the status thats coming when `this.hasConsented` is true, right? The only difference is in the way the surrounding site handles cookie storage, but that's outside of the cc code/functions anyway?

**Comment by iamwebrocker** [comment]
Argh! While browsing through the examples I noticed that there are three types of buttons involved - this explains a lot of my confusion. :-) 
We have cc-dismiss -- that's the one in the notification banner "ok, got it" - sets a status of "dismiss".
*I* "upgraded a informal banner to a opt-in one - but kept the markup, so now have a dissmiss and allow button, where in the official examples are allow and *deny* buttons!
So my code above is shaky - it should also (or exclusively) check for the allow and deny status, nut allow and dismiss. Sorry for the confusion.

**Comment by thexmanxyz** [comment]
@iamwebrocker I thinks this a bug I already submitted a PR which was already merged (#468 #464 #569)

**Comment by thexmanxyz** [comment]
@iamwebrocker So from what I understand you are right with what you are writing but I think the real origin of the confusion is the wrong layout of buttons which was introduced with `3.1.0`. Let me know if that solves your questions.

**Comment by thexmanxyz** [comment]
@iamwebrocker I totally agree with your explanation. That is how it supposed to work. But the `hasConsented(...)` is with a high chance a left over from earlier versions of the library. So you are not tied to the example code ;).

**Comment by relicmelex** [comment]
@iamwebrocker & @thexmanxyz Yeah I believe the `hasConsented` is not only misleading but certainly left over from an earlier version.

Yes the correct way of thinking about it is:
```
onInitialise: function (status) {
  assert( this.hasConsented(), "User has interacted with the banner" )
  assert( status, "User has consented to allow cookies" )
}
```

I linked this to the docs updates issue.

**Comment by TheMangoTrain** [comment]
Indeed, it's good to know whether the user has already interacted with the prompt (ie accepted or rejected) because this would change the user experience accordingly.

For example...

A - If the user has never interacted with the prompt, the bottom bar would continue to show on every page load.

B - If the user has previously interacted with the prompt (accepted or rejected), then any subsequent interaction with cookie-related links (ie clicking "Cookie Policy" or "Cookie Settings" somewhere on the page) should open a full modal (or a new page) with full information and full controls. Conversely, it would be confusing to be shown the same new-visitor bottom bar when they've already gone through that experience (a user might even question if the software is even functioning correctly if they see the same thing).

This seems to be how many cookie user experiences occur on major EU-aware corporate websites.

In case A, they could bring up the same rich modal found in case B, if they click a "More information" or a link like "Customize my settings". That modal would ideally contain certain language for the new visitor, perhaps, and they can be more educated here because there's more real estate than a bottom bar.

**Comment by bloodcrow777** [comment]
> 
> 
> @iamwebrocker & @thexmanxyz Yeah I believe the `hasConsented` is not only misleading but certainly left over from an earlier version.
> 
> Yes the correct way of thinking about it is:
> 
> ```
> onInitialise: function (status) {
>   assert( this.hasConsented(), "User has interacted with the banner" )
>   assert( status, "User has consented to allow cookies" )
> }
> ```
> 
> I linked this to the docs updates issue.

This function is still being described as usable via the main documentation on the site: https://www.osano.com/cookieconsent/documentation/javascript-api/
